datajunction-server/datajunction_server/sql/decompose.py

```diff
@@ -0,0 +1,146 @@
+"""Used for extracting measures form metric definitions."""
+from pydantic import BaseModel
+
+from datajunction_server.enum import StrEnum
+from datajunction_server.sql import functions as dj_functions
+from datajunction_server.sql.parsing.backends.antlr4 import ast, parse
+
+
+class Aggregability(StrEnum):
+    """
+    Type of allowed aggregation for a given measure.
+    """
+
+    FULL = "full"
+    LIMITED = "limited"
+    NONE = "none"
+
+
+class AggregationRule(BaseModel):
+    """
+    Measures are components used to build metrics (e.g., sales_amount, revenue, user_count).
+    """
+
+    type: Aggregability = Aggregability.NONE
+    level: list[str] | None = None
+
+
+class Measure(BaseModel):
+    """
+    Measures are components used to build metrics (e.g., sales_amount, revenue, user_count).
+    """
+
+    name: str
+    expression: str  # A SQL expression for defining the measure
+    aggregation: str
+    rule: AggregationRule
+
+
+class MeasureExtractor:
+    """
+    Extracts aggregatable measures from a metric definition and generates SQL
+    derived from those measures.
+    """
+
+    def __init__(self):
+        """Register handlers for aggregation functions"""
+        self.handlers = {
+            dj_functions.Sum: self._generic_additive_agg,
+            dj_functions.Count: self._generic_additive_agg,
+            dj_functions.Max: self._generic_additive_agg,
+            dj_functions.Min: self._generic_additive_agg,
+            dj_functions.Avg: self._avg,
+        }
+
+    def extract_measures(self, metric_query: str) -> tuple[list[Measure], ast.Query]:
+        """
+        Decomposes the metric query into its constituent aggregatable measures and
+        constructs a SQL query derived from those measures.
+        """
+        query_ast = parse(metric_query)
+        measures = []
+
+        for idx, func in enumerate(query_ast.find_all(ast.Function)):
```

Comment:
Actually these are already just aggregate functions, because at the moment they're limited to the ones registered under `self.handlers`, but let me add a check anyway to future-proof this.